runtime/common/src/purchase.rs

```diff
@@ -101,31 +82,70 @@ pub struct AccountStatus<Balance> {
 	vat: Permill,
 }
 
-decl_event!(
-	pub enum Event<T> where
-		AccountId = <T as frame_system::Config>::AccountId,
-		Balance = BalanceOf<T>,
-		BlockNumber = <T as frame_system::Config>::BlockNumber,
-	{
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// Balances Pallet
+		type Currency: Currency<Self::AccountId>;
+
+		/// Vesting Pallet
+		type VestingSchedule: VestingSchedule<Self::AccountId, Moment=Self::BlockNumber, Currency=Self::Currency>;
+
+		/// The origin allowed to set account status.
+		type ValidityOrigin: EnsureOrigin<Self::Origin>;
+
+		/// The origin allowed to make configurations to the pallet.
+		type ConfigurationOrigin: EnsureOrigin<Self::Origin>;
+
+		/// The maximum statement length for the statement users to sign when creating an account.
+		#[pallet::constant]
+		type MaxStatementLength: Get<u32>;
```

Comment:
Why is this changed into a `u32`?

Reply:
`MaxStatementLength` constant is `u32`. Changing the type would make it more clean, or we have to use `pallet::extra_constants` to manually convert the type. There is no breaking change so imho it's reasonable. (example of changing length type in [Substrate migration PR](https://github.com/paritytech/substrate/pull/8762/files#diff-e44afa3f58035adbf7d19adf68eed43df626e55f7fba287641e4a4546d104b18R147))

Reply:
Technically it is a breaking change, because `usize` on 32-bit systems is 32 bits long, but on 64-bit systems it is 64 bits long, while `u32` is fixed to be 32 bits. In other words, if the user has previously configured their `MaxStatementLength` to be larger than `u32::MAX`, they may run into issues when upgrading this pallet.

Reply:
By no breaking change I mean purchase is not used in any runtime yet. If `MaxStatementLength` is configured to be larger `u32::MAX`, users will have trouble running on 32-bit system even without type change here.

Reply:
Please do not dismiss as an ancillary concern. The issue isn't whether you can imagine cases of how it can or can't break, it's that you're changing the public API here, which IS a breaking change. I don't seem to see a good reason here other than the reluctance of using an extra attribute, unless I'm not hearing you correctly on what the difficulty is?

Reply:
If you insist, of course, in strict definition, it IS a breaking change.

As expressed in the beginning of this conversion, I think it will make code more clean. I have no idea why you think it's a reluctance of doing more work using an extra attribute, or other difficulty. If that's the case, I won't bother to spend time to migrate any pallet.

Reply:
If your cleanliness change makes a breaking API change, then I don't think it's worth it. This isn't to dismiss your work here; it's that public API changes involves a lot of other downstream changes, such as polkadot-js front-end modifications to support a different type, and others that both you and I didn't even think of at this moment in time.

Again I stress, do not dismiss public API changes as ancillary.

Reply:
Please don't assume I "dismiss public API changes as ancillary".

In which way exactly this type change will do breaks, other than in theory it's a breaking change? For PolkadotJS there is no new type introduced, and there is no metadata change since the constant was `u32`. This associated type is only used "publicly" in unit tests, if I don't miss anything, what is the "others that both you and I didn't even think of at this moment in time"?

Reply:
We're hosting this as FOSS, I don't think you can confidently say that there is absolutely 0 downstream users of this crate, especially when it's in the public domain.

Reply:
I could argue more about "should we do technically breaking change like this", or "how much time I've spent to fix reasonable breaking changes of substrate/polkadot since 2018".

There's just too many off-topic things happening here and sorry I don't have all day for this.  I'm happy to revert it if any party say this is making their life difficult. Or feel free to close this PR. That's it.

Reply:
This should be fine. This should be a u32 to begin with, and the value should never have exceeded `u32::max()`.